src/runtime/io.cpp

```diff
@@ -294,6 +295,93 @@ extern "C" LEAN_EXPORT obj_res lean_io_prim_handle_mk(b_obj_arg filename, uint8
     }
 }
 
+/* Handle.lock : (@& Handle) → (exclusive : Bool) → IO Unit */
+extern "C" LEAN_EXPORT obj_res lean_io_prim_handle_lock(b_obj_arg h, uint8_t x, obj_arg /* w */) {
+    FILE * fp = io_get_handle(h);
+#ifdef LEAN_WINDOWS
+#ifdef q4_WCE
+    HANDLE wh = (HANDLE)_fileno(fp);
+#else
+    HANDLE wh = (HANDLE)_get_osfhandle(_fileno(fp));
+#endif
+    OVERLAPPED o = {0};
+    if (LockFileEx(wh, x ? LOCKFILE_EXCLUSIVE_LOCK : 0, 0, MAXDWORD, MAXDWORD, &o)) {
+        return io_result_mk_ok(box(0));
+    } else {
+        return io_result_mk_error((sstream() << GetLastError()).str());
+    }
+#else
+    if (!flock(fileno(fp), x ? LOCK_EX : LOCK_SH)) {
+        return io_result_mk_ok(box(0));
+    } else {
+        return io_result_mk_error(decode_io_error(errno, nullptr));
+    }
+#endif
+}
```

Comment:
I'd suggest lifting the bloc inside the `#ifdef q4_WCE block` into it's own function.

I also have a preference for lifting the `#ifdef LEAN_WINDOWS` outside of `_lock`, `_try_lock` and `_unlock`.  I think that would make it easier to match up the calls to the underlying OS functions across versions.

Reply:
@joehendrix I have implemented these changes (and removed the shared option) over at #2842.